### ISSUE TITLE
Feature request: Parent method call inspection improvements

### ISSUE THREAD

**Issue by asgraf** [hammer-tools/php-hammer#5]
In CakePHP framework there are many framework classes with empty event callback methods for example:
https://github.com/cakephp/cakephp/blob/4.x/src/Controller/Controller.php#L1004-L1056

In CakePHP framework controller inheritance looks like this:
All controllers extend `\App\Controller\AppController` class, that extends `\Cake\Controller\Controller`

If i create `beforeRender` method in my `MyExampleController` i will get warning from your [Senseless call to empty parent method](https://github.com/hammer-tools/php-hammer/wiki/Inspections#senseless-call-to-empty-parent-method) inspection and it will be correct at that time.
Few weeks later i might want to create global `beforeRender` callback in via adding `AppController::beforeRender` method.
Since `MyExampleControler::beforeRender` method had parent call  removed newly added `AppController::beforeRender` method will be never called in `MyExampleController::beforeRender`

So i have two suggestions:
1) In order to complement [Senseless call to empty parent method](https://github.com/hammer-tools/php-hammer/wiki/Inspections#senseless-call-to-empty-parent-method) inspection there should be another reverse inspection that checks for missing parent call when non empty parent method exists (to help spot bugs introduced earlier)
2) [Senseless call to empty parent method](https://github.com/hammer-tools/php-hammer/wiki/Inspections#senseless-call-to-empty-parent-method) inspection should have deny list option. Preferably with wild-star support.
In case of CakePHP framework this inspection can introduce bugs on `before*` and `after*` methods

**Comment by asgraf** [comment]
Another more generic option would be to limit [Senseless call to empty parent method](https://github.com/hammer-tools/php-hammer/wiki/Inspections#senseless-call-to-empty-parent-method) inspection to classes that are direct descendants of class with empty method like `AppController` but not active in indirect descendants like in `MyExampleControler`

**Comment by rentalhost** [comment]
What you want is to ensure that `parent::method()` is declared if the parent method is not empty. That's it?

```php
class A { function x() { /** empty */ } }
class B extends A { function x() { parent::a(); } } // Can be dropped.
```

```php
class A { function x() { /** empty */ } }
class B extends A { function x() { something(); } } // Must be kept.
class C extends B { function x() { /** empty */ } } // Must call parent::x() <-- that?
```

If so, weirdly I think PhpStorm already does something like this (I will check).

Would it be possible to give a more detailed example of what it is and how you expect it to be?

**Comment by asgraf** [comment]
```php
class A {
    function methodX() {
        /** empty function */
    }
}
class B extends A {}
class C extends B {}
class D extends B {}
class E extends B {}
class F extends B {}
class G extends B {
    function methodX() {
        /** some code GGG */
        return parent::methodX();
    }
}
```
In situation above B does not have own `methodX`. It currently inherits empty `methodX` method from A.
Currently [Senseless call to empty parent method](https://github.com/hammer-tools/php-hammer/wiki/Inspections#senseless-call-to-empty-parent-method) inspection will activate on `C::methodX` and it will suggest removing line
```php
return parent::methodX();
```
Doing so will not brake code now, it will brake code much time later.

Suppose that 6 months later, due to a completely different ticket, someone makes the following Class B modification:
```php
class B extends A {
    function methodX() {
        /** some code BBB */
    }
}
```
The `some code BBB` will be called correctly on classes B,C,D,E and F but due to the removal of the parent call 6 months earlier the `some code BBB` will not be called on class G
So removing the `return parent::methodX();` in class G generated a bug 6 months later because calling methodX should call both `some code GGG` and `some code BBB` instead of calling only `some code GGG`

**Comment by asgraf** [comment]
> If so, weirdly I think PhpStorm already does something like this

PHPStorm has inspection that checks existence of parent method but it does not check if that method is empty or not.
This conflicts with your [Senseless call to empty parent method](https://github.com/hammer-tools/php-hammer/wiki/Inspections#senseless-call-to-empty-parent-method) inspection because it suggests re-adding the parent call which your inspection just removed
PHPStorm's inspection always suggest adding parent call event if parent method is empty.
And one more thing:
I you plan to implement `reverse inspection that checks for missing parent call when non empty parent method exists` i would also classify methods consisting of a single line always throwing an exception as empty and not suggest to call them
```php
function someMethod()
{
   throw new \Exception('Not implemented';)
}

**Comment by rentalhost** [comment]
Just to make sure I understand: actually you want something like "_missing parent call_" to be displayed for the method in `G`, when the method in `B` or `A` exists and is other than empty.

```php
class A {
    function methodX() { /** empty */ }
}

class B extends A {
    function methodX() { 
        /** method not implemented until v1.0, but implemented at v1.1 */ 
       doSomething(); // implemented at v1.1
        /** parent call still not required here, as A::methodX() still empty */
    }
}

class G extends B {
    function methodX() {
        parent::methodX(); /** not required until v1.0, but now it is required at v1.1 because B::methodX() is implemented not-empty */
        // ^^^^^^^^^^ warn if it is not declared here at v1.1
        /** something */
    }
}
```

**Comment by asgraf** [comment]
>Just to make sure I understand: actually you want something like "missing parent call" to be displayed for the method in G, when the method in B or A exists and is other than empty.

Yes

**Comment by asgraf** [comment]
I think it would also be useful to display a warning next to B::methodX indicating that there is a method G::methodX that does not parent call B::methodX

**Comment by rentalhost** [comment]
**Implemented to version 0.20.0:** it takes about three days to be published on the market. But if you want, you can download it from **[Releases](https://github.com/hammer-tools/php-hammer/releases)**.

**Notes:**

While I understand it's a necessary feature, the inspection won't trigger for parent methods that only `return` something. Because, in general, the override is intentional (please let me know if I'm wrong, in which case I can create a new option).

Example:

```php
class Parent {
    public function method() {
        return doSomething(); // <-- Just a return.
    }
}

class Child extends Parent {
    public function method() {
        return doAnotherThing(); // <-- Intentional override.
    }
}
```

As for your second request, I believe it is not interesting at the moment, since PhpStorm can validate the entire project in search of problems. So we would end up having two reports of the same problem: one from the child to the parent and one from the parent to the child.

**Additional notes:**

* Enabled by default;
* Support to quick-fix;
